procrustes/kopt.py

```diff
@@ -97,71 +96,70 @@ def kopt_heuristic_double(array_m, array_n, ref_error,
+    # assign p & q to be an identity arrays, if not specified
+    if p is None:
+        p = np.identity(np.shape(a)[0])
+    if q is None:
+        q = np.identity(np.shape(a)[0])
+
+    num_row_left = p.shape[0]
+    num_row_right = q.shape[0]
+    kopt_error = compute_error(a, b, p, q)
     # the left hand side permutation
     # pylint: disable=too-many-nested-blocks
-    for comb_left in it.combinations(np.arange(num_row_left), r=kopt_k):
-        for comb_perm_left in it.permutations(comb_left, r=kopt_k):
+    for comb_left in it.combinations(np.arange(num_row_left), r=k):
+        for comb_perm_left in it.permutations(comb_left, r=k):
             if comb_perm_left != comb_left:
-                perm_kopt_left = deepcopy(perm_p)
+                perm_kopt_left = deepcopy(p)
                 # the right hand side permutation
-                for comb_right in it.combinations(np.arange(num_row_right), r=kopt_k):
-                    for comb_perm_right in it.permutations(comb_right, r=kopt_k):
+                for comb_right in it.combinations(np.arange(num_row_right), r=k):
+                    for comb_perm_right in it.permutations(comb_right, r=k):
                         if comb_perm_right != comb_right:
-                            perm_kopt_right = deepcopy(perm_q)
+                            perm_kopt_right = deepcopy(q)
                             perm_kopt_right[comb_right, :] = perm_kopt_right[comb_perm_right, :]
-                            e_kopt_new_right = compute_error(array_n, array_m, perm_p.T,
+                            e_kopt_new_right = compute_error(b, a, p.T,
                                                              perm_kopt_right)
                             if e_kopt_new_right < kopt_error:
-                                perm_q = perm_kopt_right
+                                q = perm_kopt_right
                                 kopt_error = e_kopt_new_right
-                                if kopt_error <= kopt_tol:
+                                # check whether perfect permutation matrix is found
+                                # TODO: smarter threshold based on norm of matrix
+                                if kopt_error <= 1.0e-8:
                                     break
 
                 perm_kopt_left[comb_left, :] = perm_kopt_left[comb_perm_left, :]
-                e_kopt_new_left = compute_error(array_n, array_m, perm_kopt_left.T, perm_q)
+                e_kopt_new_left = compute_error(b, a, perm_kopt_left.T, q)
                 if e_kopt_new_left < kopt_error:
-                    perm_p = perm_kopt_left
+                    p = perm_kopt_left
                     kopt_error = e_kopt_new_left
-                    if kopt_error <= kopt_tol:
+                    # check whether perfect permutation matrix is found
+                    # TODO: smarter threshold based on norm of matrix
+                    if kopt_error <= 1.0e-8:
                         break
```

Comment:
I am a bit confused regarding the permutations tried in the `kopt_heuristic_double`. It seems to me that the swaps of right-hand-side permutation matrix P and left-hand-side permutation matrix Q are tested separately (to check whether it lowers the error). What about swapping both P & Q, and checking how the error would change?

Reply:
They should be tested together. You want to generate *all* **P** and **Q** and then check for all possible combinations.

Reply:
Ok, it is addressed in a1cfd3b.